Applications/Spire/Source/Playback/PlaybackControlPanel.cpp

```diff
@@ -0,0 +1,419 @@
+  auto controls_box = new Box(controls_body);
+  layout->addWidget(controls_box);
+  layout->addStretch();
+  auto box = new Box(body);
+  proxy_style(*this, *box);
+  enclose(*this, *box);
+  update_style(*this, [] (auto& style) {
+    style.get(Any()).
+      set(horizontal_padding(scale_width(8))).
+      set(vertical_padding(scale_height(8)));
+  });
+  on_playhead_update(get_playhead()->get());
+  on_state_update(m_state.get());
+  m_state.connect_update_signal(
+    std::bind_front(&PlaybackControlPanel::on_state_update, this));
+  m_date_connection = date_box->connect_submit_signal(
+    std::bind_front(&PlaybackControlPanel::on_date_submit, this));
+  m_playhead_connection = m_playhead->connect_update_signal(
+    std::bind_front(&PlaybackControlPanel::on_playhead_update, this));
+  m_timeline_connection = get_timeline()->connect_update_signal(
+    std::bind_front(&PlaybackControlPanel::on_timeline_update, this));
+}
+
+const std::shared_ptr<TimelineModel>&
+    PlaybackControlPanel::get_timeline() const {
+  return m_playhead->m_timeline;
+}
+
+TimeClientBox PlaybackControlPanel::get_time_client() const {
+  return m_playhead->m_time_client;
+}
+
+const std::shared_ptr<DurationModel>&
+    PlaybackControlPanel::get_playhead() const {
+  return m_playhead->m_source;
+}
+
+const std::shared_ptr<PlaybackSpeedModel>&
+    PlaybackControlPanel::get_speed() const {
+  return m_speed_box->get_current();
+}
+
+connection PlaybackControlPanel::connect_state_signal(
+    const StateSignal::slot_type& slot) const {
+  return m_state.connect_update_signal(slot);
+}
+
+void PlaybackControlPanel::keyPressEvent(QKeyEvent* event) {
+  if(event->key() == Qt::Key_Space && m_seek_bar->hasFocus() &&
+      m_play_toggle->isEnabled()) {
+    m_play_toggle->get_current()->set(!m_play_toggle->get_current()->get());
+    event->accept();
+  } else if(event->key() == Qt::Key_Home && m_start_button->isEnabled()) {
+    on_jump_to_start_click();
+    event->accept();
+  } else if(event->key() == Qt::Key_End && m_end_button->isEnabled()) {
+    on_jump_to_end_click();
+    event->accept();
+  } else {
+    QWidget::keyPressEvent(event);
+  }
+}
+
+bool PlaybackControlPanel::is_playhead_at_end() const {
+  return get_playhead()->get() >= get_timeline()->get().m_duration;
+}
+
+void PlaybackControlPanel::play() {
+  if(get_timeline()->get().m_start + m_playhead->get() <
+      m_playhead->m_time_client.GetTime() - TIME_TOLERANCE) {
+    if(m_state.get() != State::REPLAYING) {
+      m_state.set(State::REPLAYING);
+    }
+  } else if(m_state.get() != State::REAL_TIME) {
+    m_state.set(State::REAL_TIME);
+  }
+}
+
+void PlaybackControlPanel::pause() {
+  if(m_state.get() != State::PAUSED) {
+    m_state.set(State::PAUSED);
+  }
+}
+
+void PlaybackControlPanel::on_play_checked(bool checked) {
+  auto& tooltip = *m_play_toggle->findChild<Tooltip*>("Tooltip");
+  if(checked) {
+    tooltip.set_label(tr("Pause (Space)"));
```

Comment:
Why is the label changed here?

Reply:
`IconToggleButton` doesn’t support switching tooltips based on its checked state by default. The code here does the job. Are you suggesting adding a new API to allow `IconToggleButton` to accept different tooltips for each check state?